frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -262,6 +234,86 @@ function DataFrame({
   const { columns, sortColumn, getOriginalIndex, getCellContent } =
     useColumnSort(originalNumRows, originalColumns, getOriginalCellContent)
 
+  const applySelections = React.useCallback(
+    // Use debounce to prevent rapid updates to the widget state.
+    debounce(DEBOUNCE_TIME_MS, (newSelection: GridSelection) => {
+      // If we want to support selections also with in the editable mode,
+      // we would need to integrate the `applyEdits` and `applySelections` functions
+      // into a single function that updates the widget state with both the editing
+      // state and the selection state.
+
+      const selectionState = {
+        // We use snake case here since this is the widget state
+        // that is sent and used in the backend. Therefore, it should
+        // conform with the Python naming conventions.
+        select: {
+          rows: [] as number[],
+          columns: [] as string[],
+        },
+      }
+
+      selectionState.select.rows = newSelection.rows.toArray().map(row => {
+        return getOriginalIndex(row)
+      })
+      selectionState.select.columns = newSelection.columns
+        .toArray()
+        .map(columnIdx => {
+          return getColumnName(originalColumns[columnIdx])
+        })
+      const newWidgetState = JSON.stringify(selectionState)
```

Comment:
suggestion: I would move this part to a function outside of the React component, as it does not use anything of it.

```react
function getSelectionState(newSelection: GridSelection, getIndexForRow: (row: Row) => index, getColumnName: (column: Column) => string): GridSelection {
        const selectionState = {
        // We use snake case here since this is the widget state
        // that is sent and used in the backend. Therefore, it should
        // conform with the Python naming conventions.
        select: {
          rows: [] as number[],
          columns: [] as string[],
        },
      }

      selectionState.select.rows = newSelection.rows.toArray().map(row => {
        return getOriginalIndex(row)
      })
      selectionState.select.columns = newSelection.columns
        .toArray()
        .map(columnIdx => {
          return getColumnName(originalColumns[columnIdx])
        })
      return selectionState;
}
```

and then call that function from here and passing the component-dependent information. This way only the component-specific parts are in here and the logic can live standalone. In general this should make it easier to understand a components-flow by reducing the number of lines within a component and clearly seeing where an update is triggered.

Reply:
I will eventually refactor this into a dedicated `useWidgetState` custom hook. But doing the refactor for chart selections is a bit too much overhead and risk of breaking something. So I think its good to leave it as is and do some cleanup after the deadlines. There is anyways a lot to touche and refactor in the context of the lazy-loading support project.